### ISSUE TITLE
[BUG] VMs from successive "Restore New" backup/snapshot tend to get the same IP. 

### ISSUE THREAD

**Issue by albinsun** [harvester/harvester#4474]
## Description
VMs from successive "Restore New" backup/snapshot tend to get the same IP even they have different MAC.
![image](https://github.com/harvester/harvester/assets/2773781/7e2a77b3-3cf9-4b69-bedb-a57e87cd0aff)

## Environment
* Harvester
  * Version: `v1.2.0-rc6`
  * Profile: QEMU/KVM, single node (8C/16G/250G)
  * **ui-source**: Auto

## To Reproduce
1. Create image `focal-server-cloudimg-amd64.img`
2. Create VLAN1 network `mgmt-vlan1`
3. Create 2 VMs, wait them Running and get IP
   * vm-ctrl as control group
   * vm-exp as experiment group
4. Create snapshot `vm-exp-snap` for `vm-exp`, check it's Active
5. Go to "Backup & Snapshot" -> "VM Snapshots" -> `vm-exp-snap`
6. "Restore New" twice as `vm-exp-snap1` and `vm-exp-snap2`
7. Wait `vm-exp-snap1` and `vm-exp-snap2` Running and get IP
8. They should get different IP since they are different MAC ${~~~\color{red}\textsf{X, they receive the same IP}}$

[duplicated_ip.webm](https://github.com/harvester/harvester/assets/2773781/9b111afc-27c9-480d-b5dc-6f136cb7b52f)

## Expected Behavior
VM via "Restore New" has its own MAC, thus should NOT get the same IP

## Note
1. From current test, this issue only happens in VM using VLAN network.
2. Both backup & snapshot restore has this issue.
3. Note that due to https://github.com/harvester/harvester/issues/4473, the original VM `vm-exp` will also have IP change later.
![Screenshot_20230831_230824](https://github.com/harvester/harvester/assets/2773781/511759a2-9a73-4860-a02a-b63b39c52cc5)

## Workaround?
1. Reboot can NOT workaround, will get the same IP.
4. Start 2nd restore after 1st done has chance to get different IP as expected.

## Support Bundle
[supportbundle_ab1df3fb-9fa8-40ec-b926-de1e41c86c97_2023-08-31T15-15-07Z.zip](https://github.com/harvester/harvester/files/12488647/supportbundle_ab1df3fb-9fa8-40ec-b926-de1e41c86c97_2023-08-31T15-15-07Z.zip)

## Related Ticket
1. https://github.com/harvester/harvester/issues/4473

**Comment by guangbochen** [comment]
I can reproduce this case in both v1.1.2 and v1.2.0. However, this is a known limitation of cloud images like Ubuntu since its `machine-id` will be identical when restoring VMs from the snapshot/backup; you may refer to https://www.ubackup.com/enterprise-backup/cloned-vm-has-same-ip.html for more details.
As a workaround, you can regenerate a machine-id of the restored VM by:
```shell
rm -f /etc/machine-id
dbus-uuidgen --ensure=/etc/machine-id
```
Then, reboot the VM to get a new IP via DHCP.

For further enhancement, we may consider allowing users to configure the `user-data` when restoring a new VM, therefore, they can add the above workaround to the cloud-init script.

**Comment by mingshuoqiu** [comment]
The dhcp client will use /etc/machine-id by default for getting IP lease from server. Or we can add `send dhcp-client-identifier hw;` in `/etc/dhcp/dhclient.conf` to force the dhcp client to use HW mac as identifier for getting IP.
I tried adding this line in dhclient.conf before take snapshot, then `Restore New` with it. Then the problem won't happen since the mac address of `vm-exp` and `vm-exp-snap` are different.

If I remove `/etc/machine-id` before snapshot, then the restored VM will fail to boot since the /etc/machine-id is only generated in firstboot. User have to manually replace the machine-id with new one and reboot to make it work. Unless we can force regenerate the machine-id for the `cloned` VM before running it, it's hard to fix the problem.

**Comment by bk201** [comment]
@mingshuoqiu Can you try Ubuntu jammy image? I seem to see this problem with the focal image, but not in jammy.

**Comment by mingshuoqiu** [comment]
Actually I can reproduce with jammy image. The problem is majorly caused by the same `machine-id` of the snapshoted machine. To fix this issue, we can only either generate different machine-id for the snapshot'd VM or force the dhcp client to use `hw` mac as dhcp identifier when doing snapshot/backup.

**Comment by bk201** [comment]
@mingshuoqiu This happened in https://github.com/harvester/harvester/issues/6402, can you help figure out a good practice for users to do before creating a backup?

**Comment by mingshuoqiu** [comment]
@albinsun Can we access the configuration file on the DHCP server for #6402 ? I'd like to suggest change on dhcp server config instead of the VM image.

**Comment by mingshuoqiu** [comment]
We can provide a template network data for cloud config which has the `dhcp-identifier: mac` configured. Ubuntu use netplan since 18.04 (or even earlier 16.04) for network configuration and the `/etc/netplan/50-cloud-init.yaml` is for digesting the cloud init config.(https://cloudinit.readthedocs.io/en/latest/reference/network-config-format-v2.html)
```
network:
    ethernets:
        enp1s0:
            dhcp4: true
            dhcp-identifier: mac
    version: 2
```

I tried creating VM w/ Ubuntu image and the above network data for cloud config. The problem no longer exist for the snapshot/restore.

<img width="1182" alt="截圖 2024-09-02 下午5 08 50" src="https://github.com/user-attachments/assets/3048b48f-51c3-4c5a-b296-e348fd89bd24">
We can have the default network data to enable the default dhcp client id in cloud config for the `OS Type` Ubuntu. How about it?

**Comment by bk201** [comment]
@albinsun can you help verify if the suggestion by Chris works well? We can then add a doc and update the template.

**Comment by bk201** [comment]
@mingshuoqiu Is this a doc fix?

**Comment by mingshuoqiu** [comment]
Which doc should I update? And can I provide the template directly on the doc?

**Comment by mingshuoqiu** [comment]
> @mingshuoqiu Is this a doc fix?

As I know, we should provide the template for VM which installed with UBUNTU image. The problem is we still need to provide a reminder that user have to modify the config per their VM network configuration

**Comment by bk201** [comment]
> Which doc should I update? And can I provide the template directly on the doc?

I guess we can add the comment to the doc:
https://github.com/harvester/harvester/issues/4474#issuecomment-2324226625

**Comment by albinsun** [comment]
> [@albinsun](https://github.com/albinsun) can you help verify if the suggestion by Chris works well? We can then add a doc and update the template.

Sorry I missed the message and reply late.😛 
Just tried @mingshuoqiu's suggestion, looks good and does not hit the issue with **Network Data**:
```
network:
    ethernets:
        enp1s0:
            dhcp4: true
            dhcp-identifier: mac
    version: 2
```

Note that we should also evaluate this is a solution or workaround cuz
1. The NIC is fixed on _enp1s0_, what if this VM has multiple NICs?
2. Configure on DHCP4, how about DHCP6, are there critical side effects? 


![Image](https://github.com/user-attachments/assets/5d11cb07-d9fc-4ea2-afa5-18a330176104)
* ubuntu-template-current
  ![Image](https://github.com/user-attachments/assets/d52b01e2-a1b3-4c37-a9a8-10fee75e4aca)

* ubuntu-template-netdata 
  ![Image](https://github.com/user-attachments/assets/bcbde708-30a7-4052-a85a-79bcf414f8cb)

**Comment by mingshuoqiu** [comment]
Agree. So it would be a template for user's reference.  It needs to pop up a warning message that user needs to modify this according to the underlying network configurations.

The  client-identifier is the same effect for DHCPv4 and DHCPv6.

**Comment by mingshuoqiu** [comment]
Just confirmed that the cloud init data for different os type is hardcoded at.
https://github.com/harvester/harvester-ui-extension/blob/f9a277d893d65e473b4c5038132e6b96a7eceba1/pkg/harvester/mixins/harvester-vm/impl.js#L11-L22

I believe the netdata can also be written here for Ubuntu.

For 1.5.0, I think we can provide document for user's reference first with the existing VMs and creating VM.

**Comment by mingshuoqiu** [comment]
Closed since the document has been merged.
